File: solutions/python/protein-translation/1/protein_translation.py
""" Module to translate DNA codon to proteins"""
def proteins(strand):
    """ Function to translate DNA codon to proteins"""
    c_to_a = {"AUG":"Methionine", "UUU":"Phenylalanine", "UUC":"Phenylalanine",
              "UUA":"Leucine", "UUG":"Leucine", "UCU":"Serine", "UCC":"Serine",
              "UCA":"Serine", "UCG":"Serine", "UAU":"Tyrosine", "UAC":"Tyrosine",
              "UGU":"Cysteine", "UGC":"Cysteine", "UGG":"Tryptophan",
              "UAA":"STOP", "UAG":"STOP", "UGA":"STOP",}
    i = 0
    protein_list = []
    codon = ""
    while (i < len(strand) and codon not in ["UAA", "UAG", "UGA"]):
        codon = strand[i:i+3]
        if codon not in ["UAA", "UAG", "UGA"]:
            protein_list.append(c_to_a[codon])
        i += 3

    return protein_list

File: solutions/javascript/protein-translation/1/protein-translation.js
//
// This is only a SKELETON file for the 'Protein Translation' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const translate = (codons) => {
  var protein=[];
  if (!codons) return protein;
  var trans = {'AUG':'Methionine',
  'UUU':'Phenylalanine', 'UUC':'Phenylalanine',
  'UUA':'Leucine', 'UUG':'Leucine',
  'UCU':'Serine', 'UCC':'Serine', 'UCA':'Serine', 'UCG':'Serine',
  'UAU':'Tyrosine', 'UAC':'Tyrosine',
  'UGU':'Cysteine', 'UGC':'Cysteine',
  'UGG':'Tryptophan',
  'UAA':'STOP', 'UAG':'STOP', 'UGA':'STOP'};
  for (let i = 0; 3*i<codons.length; i++){
    var codon = codons.substring(3*i,3*i+3);
    if (codon in trans) {
      if (trans[codon] == 'STOP') return protein;
      protein.push(trans[codon])
    } else {
      throw new Error('Invalid codon');
    }
  }
  return protein;
};
